### ISSUE TITLE
Fully Support GenT in API

### ISSUE THREAD

**Issue by EarthCitizen** [hedgehogqa/haskell-hedgehog#208]
Sometimes it is necessary to embed something in `GenT` other than `Identity`. `GenT` is exported, but functions such as `sample` and `forAll` expect `Gen`. Please add full support in the API for `GenT` to be able to use it how `Gen` is used today.

**Comment by moodmosaic** [comment]
From https://github.com/hedgehogqa/haskell-hedgehog/pull/206#issuecomment-435399322 by @moodmosaic:

>See: https://stackoverflow.com/q/53106042/467754
>
>I think `forAllT` and `forAllWithT` should be exposed back when @jystic did >https://github.com/hedgehogqa/haskell-hedgehog/pull/99. /cc @thumphries 
>
>---
>
>Perhaps we should ask @EarthCitizen to reopen this and get it merged.
>

From https://github.com/hedgehogqa/haskell-hedgehog/pull/206#issuecomment-435702377 by @EarthCitizen:

>@moodmosaic I cancelled this because of realizing that simply exposing `forAllT` and `forAllWithT` is not >enough. To take full advantage of using `GenT`, many other functions have to change as well.
>
>For example `forAllT` is defined as:
>```
>forAllT :: (Monad m, Show a, HasCallStack) => GenT m a -> PropertyT m a
>```
>
>But `property` is defined as:
>```
>property :: HasCallStack => PropertyT IO () -> Property
>```
>If you wanted to have `GenT State a` from `forAllT`, you would not be able to pass it to `property`.
>
>And `sample` is defined as:
>```
>sample :: MonadIO m => Gen a -> m a
>```
>
>Overall, there are many changes that need to occur to allow a more "open" use of `GenT`.
>
>I am guessing that these functions should be:
>```
>property :: (HasCallStack, MonadIO m) => PropertyT m () -> Property
>sample :: MonadIO m => GenT m a -> m a
>```
>
>And perhaps it would be better if `forAll` and `forAllWith` were _**something**_ like this:
>```
>forAll :: (Monad m, Show a, HasCallStack) => GenT m a -> PropertyT m a
>forAllWith :: (Monad m, HasCallStack) => (a -> String) -> GenT m a -> PropertyT m a
>```
>These suggestions are likely not exactly what is needed, but rather to demonstrate the point that the API >needs an overhaul in order to be able to use `GenT` seamlessly.
>
>I opened an issue for exactly this: #208

**Comment by moodmosaic** [comment]
I still think it'll be convenient to expose `forAllT` and `forAllWithT`.

As for the proposed `property` change, I'm not sure what you want to do with `property` is possible(?) (e.g. Why not a `StateT Gen` instead?)

---

It'd be easier with some concrete examples. Right now we've [received](https://github.com/hedgehogqa/haskell-hedgehog/issues/203) [some](https://github.com/hedgehogqa/haskell-hedgehog/pull/206) [requests](https://stackoverflow.com/q/53106042/467754) for this, but no examples. /cc @i-am-the-slime @EarthCitizen @dnadales

**Comment by EarthCitizen** [comment]
@moodmosaic For an example, what I am trying to achieve (which may be possible to solve with existing APIs) is that I have to test an expression evaluator. I have achieved generating (with Hedgehog) expressions of any depth for the normal kinds of operations like add, subtract, etc., but function calls are also part of my AST. In order to randomly generate the function calls, I also need to generate the function definitions and keep them in a symbol table in `State` containing the function names mapped to the function definitions.

I suppose, with a lot of effort and ugly code, somehow pass around my symbol table too all of my gen functions, but that is exactly the problem that `State` is suited to solve.

Perhaps it is possible to do this with existing APIs, but that I just haven't figured it out yet. I am by no means an expert on monad transformer stacks.

**Comment by thumphries** [comment]
Sorry for the slow response.

I am totally on board for exposing `forAllT` and `forAllWithT`. They are essential for some of my favourite Hedgehog use cases.

The type of `property` doesn't matter much, as long as you can pass some natural transformation `(forall m. m a -> IO a)` in at the top of your property.

Currently the only way to do this is to use the `MFunctor` instance (`property . hoist f`). It would be fairly reasonable to expose a `propertyT` or `hoistProperty` function to make this more discoverable (though we will most certainly bikeshed the function name to death).

I'm not quite sure what to do about `sample`. It only exists as convenient shorthand for use in GHCi, and is not as useful as it could be. We could just expose a generalized version in the short term so it is usable with GenT, fixing the acute issue, and leave the redesign for when Jake is back.

**Comment by thumphries** [comment]
As for examples, `GenT (ReaderT Env)` is incredibly useful for generating valid programs. Maybe we should add one to `hedgehog-example`.

**Comment by EarthCitizen** [comment]
@thumphries Yes, please add an example of that. Having a concrete example would be incredibly useful for this conversation and likely clarify what changes are and are not needed to APIs.

**Comment by dnadales** [comment]
> It'd be easier with some concrete examples. Right now we've received some requests for this, but no examples. /cc @i-am-the-slime @EarthCitizen @dnadales

Thanks for the reply! In the meantime I solved my original problem in some other way. Some time ago I used the `GenT` transformer of `QuickCheck` for generating [ADT's for a DSL](https://github.com/TorXakis/TorXakis/blob/issue/565-newADTstructure/sys/sort-gen/src/Sort/FieldDefsGen.hs). The idea was to generate ADT declarations. This required the generators to have access to a state, so that for instance when generating a field name they would know which names where already generated or which other ADT's were generated and could be used as field types.

So basically, every time a generator depends on some state I see myself wanting to use a `GenT (State s)` monad. But maybe there are other ways to solve this.

**Comment by EarthCitizen** [comment]
@dnadales Our needs sound very similar.

**Comment by dnadales** [comment]
> @dnadales Our needs sound very similar.

Oh, good to hear.

Regarding my original problem (generating traces, given some initial state), I solved it by passing the state [as an extra parameter](https://github.com/dnadales/sandbox/blob/280dc074f79ead31cda6cf9948488e4b4eb6ad24/sos-to-generators/src/Control/State/TransitionGen.hs#L61), so maybe this could be an alternative to the `GenT` transformer (I even coded failure to generate a value by using `MaybeT Gen`). The only question in that case is why `GenT` exists at all.

**Comment by EarthCitizen** [comment]
> As for examples, `GenT (ReaderT Env)` is incredibly useful for generating valid programs. Maybe we should add one to `hedgehog-example`.

@thumphries Any chance of that example happening soon?

**Comment by EarthCitizen** [comment]
@thumphries Revisiting what @moodmosaic had suggested (`StateT Gen a`) instead of (`GenT State a`) I was able to solve my problem of having artifact collection across the network of `MonadGen` functions. The only caveat is that it needs to be unpacked before calling `forAll`. I seems my thinking was backwards on the stack order (?).

@dnadales I believe your point about if `GenT` is needed at all is worth further discussion as I came to a similar solution to yours (`MaybeT Gen`).

With this working, this issue is no longer needed by me. It seems that my understanding of the problem was not yet complete. Perhaps the example of `GenT (ReaderT Env)` suggested by @thumphries will definitively answer the question as to whether this issue is a valid request or not.

**Comment by moodmosaic** [comment]
>(StateT Gen a) instead of (GenT State a) was able to solve my problem

@EarthCitizen, let's close this then. We can always come back to it and reopen it, or add comments to it.

Thanks for all the great feedback, all!

---

As @thumphries said, we may add something to `hedgehog-example`, and until then, we can open an issue for it, like https://github.com/hedgehogqa/haskell-hedgehog/issues/228, so we don't forget.

**Comment by EarthCitizen** [comment]
Closing. Not currently needed.
